### PR TITLE
Idempotency key for transaction API

### PR DESCRIPTION
**Pull request by korycins**
I want to merge this change because it should solve https://github.com/saleor/saleor/issues/14926

> [!IMPORTANT]  
> This PR introduces the logic for idempotency-key. Some migrations are introduced in #14991 

The implementation propagates the app.identifier with app.id when app.identifier is None (local apps). It was required to apply the unique constraint on `TransactionItem`. The unique constraint is also applied to `TransactionEvent`. 

The introduced constraints are used to block the possibility of recording the same transaction twice (in case of receiving the same requests twice in a short period of time). 
The implementation also doesn't allow to provide the idempotency key as an empty string. The mutation will raise an exception. 
In case of finding the transaction with the same idempotency key, the existing one will be used instead of creating a new one

The PR relays on changes from: https://github.com/saleor/saleor/pull/14991

The logic for idempotency-key is separated into multiple PRs to cover the zero-downtime approach:
PR 14991: Add idempotency fields
PR 14991: Add unique indexes concurrently
(this one) PR 14948: Convert indexes into constraints
(this one) PR 14948: Add logic that uses the logic

# Impact

- [x] New migrations
- [x] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [x] Link to documentation: https://github.com/saleor/saleor-docs/pull/1053

# Pull Request Checklist

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases
- [ ] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [ ] All migrations have proper dependencies
- [ ] All indexes are added concurrently in migrations
- [ ] All RunSql and RunPython migrations have revert option defined

### CODE REVIEW

File: saleor/app/migrations/tasks/saleor3_14.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+@celeryconf.app.task
+def set_local_app_identifier_in_transaction_item_task(app_ids=None):
+    if not app_ids:
+        return
+    apps = App.objects.filter(id__in=app_ids)
+    transaction_item_ids = TransactionItem.objects.filter(
+        Exists(apps.filter(id=OuterRef("app_id"))), app_identifier__isnull=True
+    ).values_list("pk", flat=True)[:TRANSACTION_ITEM_BATCH_SIZE]
+    if transaction_item_ids:
+        transaction_items = TransactionItem.objects.filter(id__in=transaction_item_ids)
```
Comment (tomaszszymanski129) [inline saleor/app/migrations/tasks/saleor3_14.py:40]: same, we could just operate on objects, one query less.

File: saleor/app/migrations/tasks/saleor3_14.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+TRANSACTION_ITEM_BATCH_SIZE = 1000
+
+
+@celeryconf.app.task
+def set_identifier_for_local_apps_task():
+    apps_ids = App.objects.filter(identifier__isnull=True).values_list("pk", flat=True)[
+        :APP_BATCH_SIZE
+    ]
+    if apps_ids:
+        local_apps = App.objects.filter(id__in=apps_ids, identifier__isnull=True)
```
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/app/migrations/tasks/saleor3_14.py:22]: Yeah, you're right. My bad, it is something that left when I was cleaning up the logic 😅  Thanks for catching this!

File: saleor/payment/migrations/0054_add_idempotency_key_contraints.py
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/payment/migrations/0054_add_idempotency_key_contraints.py:1]: After your comment, I discovered that I had an incorrect approach here. Thanks for this. 
So I slightly changed the approach.
PR: https://github.com/saleor/saleor/pull/14991/files 
Introduces the new fields for TransactionItem and TransactionEvent.  There is also migration for introducing indexes concurrently. 

The indexes are converted into constraints in this PR.

File: saleor/app/migrations/tasks/saleor3_14.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/app/migrations/tasks/saleor3_14.py]: Fixed. Thanks again

File: saleor/app/migrations/tasks/saleor3_14.py (hunk 1 above)
Comment (korycins) [reply saleor/app/migrations/tasks/saleor3_14.py]: Fixed. Thanks

**Review by michal-macioszczyk — APPROVED** [review]

**Review by szczecha — APPROVED** [review]

**Pull request merged**